src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1024,6 +930,18 @@ MasterIrreducibleExpressionWalker(Node *expression, WalkerState *state)
 }
 
 
+static bool
+MasterIrreducibleExpressionFunctionChecker(Oid func_id, void *context)
```

Comment:
volatileContext is a tri-state variable.
1. volatile
2. stable 
3. else (immutable)

Walker functions checks (1) if it has volatile function, (2) if it has stable function. Only a boolean value is not enough for this comparison.

The caller ```check_functions_in_node()``` does not call this function again once it returns true.  So the context  variable can not be volatile ever. We can use this value to indicate if it has a stable function or not. 

So the fix would be replacing ```volatileFlag``` in walker function (```MasterIrreducibleExpressionWalker```) with something like ```mostPermissiveFuntionIsStable```

I did not like this approach though.